docs/modules/schema-validation/con-record-validation.adoc

```diff
@@ -0,0 +1,83 @@
+// file included in the following:
+//
+// index.adoc
+
+[id='con-record-validation-{context}']
+= Record Validation
+
+[role="_abstract"]
+The Record Validation filter validates records sent by a producer.   Only records that pass the validation are sent to
+the broker. This filter can be used to prevent poison messages entering the Kafka system, that may otherwise lead to
+consumer failure.
+
+The filter currently supports two modes of operation:
+
+1. Schema validation validates the content of the record against a schema stored in an https://www.apicur.io/registry/[Apicurio Registry].
+2. Syntactically Correct JSON validation ensures the producer is producing messages that contain syntactically valid JSON.
+
+Validation rules can validate the content of the Kafka record key or value.
+
+If the validation fails, the product request is rejected and the producing application receives an error response.  The broker
+will not receive the rejected records.
+
+== How to use the filter
+
+There is just one step to use the filter.
+
+1. <<Configuring virtual clusters>>
+2. Configuring the filter within Kroxylicious.
+
+=== Configuring the filter within Kroxylicious.
+
+.Example Record Validation filter configuration
+
+[source,yaml]
+----
+filters:
+  - type: ProduceValidationFilterFactory
+    config:
+        rules:
+        - topicNames:                                            # <1>
+            - <topic name>
+            keyRule:
+              <rule definition>                                  # <2>
+            valueRule:
+              <rule definition>                                  # <3>
+        isForwardPartialRequests: false                          # <4>
+----
+<1> List of topic names to which the validation rules will be applied.
+<2> Validation rules that are applied to the record's key.
+<3> Validation rules that are applied to the record's value.
+<4> If false, any record that fails the validation will cause
+    the entire batch to be rejected.  If true, only records that fail validation
+    will be rejected. Defaults to false.
```

Comment:
This is wordy to explain. ProduceRequests have their data organised into TopicPartitions and the Kafka API allows  each TopicPartition write to succeed or fail independently. So this is the granularity that the proxy can choose to act on, forwarding valid TopicPartitions and blocking TopicPartitions that contain any invalid values.

The decision is: "should the proxy forward on TopicPartitions containing only valid records, despite other TopicPartitions in the produce request containing invalid records".

Maybe the ProduceRequest is carrying data from numerous topics, a reasonable client configuration to use less connections/requests. Messages for topic A could be invalid, but we could still let valid messages for topic B make it to the broker.

Then there's also the subtlety that this setting does nothing if transactions are being used.  We make the whole  request fail as a unit.

Reply:
might be a case of a picture telling a thousand words and a diagram would help

Reply:
I added some words.  Do you think this is good enough?

Reply:
yep LGTM